tests/test_recipe.py

```diff
@@ -112,6 +112,77 @@ def test_standard_recipe_with_balancer(self):
 
         self.assertEqual(counts["entailment"], counts["not entailment"])
 
+    def test_standard_recipe_with_loader_limit(self):
+        recipe = StandardRecipeWithIndexes(
+            card="cards.wnli",
+            instruction="instructions.models.llama",
+            template="templates.key_val",
+            format="formats.user_agent",
+            demos_pool_size=5,
+            num_demos=1,
+            loader_limit=10,
+        )
+
+        stream = recipe()
+        self.assertEqual(len(list(stream["train"])), 5)
+        self.assertEqual(len(list(stream["test"])), 10)
+
+    def test_standard_recipe_with_loader_limit_errors(self):
+        with self.assertRaises(ValueError):
+            recipe = StandardRecipeWithIndexes(
+                card="cards.wnli",
+                template="templates.key_val",
+                max_test_instances=10,
+                loader_limit=9,
+            )
+
+        with self.assertRaises(ValueError):
+            recipe = StandardRecipeWithIndexes(
+                card="cards.wnli",
+                template="templates.key_val",
+                max_train_instances=10,
+                loader_limit=9,
+            )
+        with self.assertRaises(ValueError):
+            recipe = StandardRecipeWithIndexes(
+                template="templates.key_val",
+                card="cards.wnli",
+                max_validation_instances=10,
+                loader_limit=9,
+            )
+
+        with self.assertRaises(ValueError):
+            recipe = StandardRecipeWithIndexes(
+                template="templates.key_val",
+                card="cards.wnli",
+                num_demos=3,
+                demos_pool_size=10,
+                loader_limit=9,
+            )
+
+    def test_standard_recipe_with_template_errors(self):
+        with self.assertRaises(AssertionError) as cm:
+            recipe = StandardRecipeWithIndexes(card="cards.wnli")
+        self.assertEqual(str(cm.exception), "Specify either template or template_card_index in card")
+
+        with self.assertRaises(AssertionError) as cm:
+            recipe = StandardRecipeWithIndexes(
+                card="cards.wnli", template="templates.key_val", template_card_index=100
+            )
+        self.assertTrue("Specify either template" in str(cm.exception))
```

Comment:
you are missing " or template_card_index in card" I think , and the assert is weird, why 2?

Reply:
maybe a better idea is to add a short description to each test, to explain the scenario

Reply:
Changed to regex search and added short description